lockfile/lockfile.go

```diff
@@ -0,0 +1,75 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lockfile
+
+import (
+	"errors"
+	sys "syscall" // XXX: if we are only targeting 15.04+, i.e. golang 1.5+, we should move to golang.org/x/sys
+)
+
+// ErrAlreadyLocked is returned when an attempts is made to lock an
+// already-locked FileLock.
+var ErrAlreadyLocked = errors.New("another snappy is running, try again later")
+
+// LockedFile is a handle you can use to Unlock a file lock.
+type LockedFile int
+
+// Lock opens (and possibly creates) a new file at the given path
+// and applies an exclusive advisory lock on it.
+//
+// If the file already has an advisory lock, the `wait` flag
+// determins whether to wait until it is removed or return
+// ErrAlreadyLocked.
+func Lock(path string, wait bool) (LockedFile, error) {
+	fd, err := sys.Open(path, sys.O_CREAT|sys.O_WRONLY, 0600)
+	if err != nil {
+		return -1, err
+	}
+
+	how := sys.LOCK_EX
+	if !wait {
+		how |= sys.LOCK_NB
+	}
+
+	if err := sys.Flock(fd, how); err == sys.EWOULDBLOCK {
+		return -1, ErrAlreadyLocked
+	} else if err != nil {
+		return -1, err
+	}
+
+	return LockedFile(fd), nil
+}
+
+// Unlock closes the file, thus also removing the advisary lock on it.
+func (fd LockedFile) Unlock() error {
+	// closing releases the lock
+	return sys.Close(int(fd))
```

Comment:
I suggest testing for a bad fd here, and resetting the fd after that to prevent double locking (will need a pointer receiver).

Using zero as the invalid fd would also mean this would catch bogus usage on unlocked files, and would perhaps be okay as zero is generally stdin.